### DISCUSSION TITLE
Workflow System For PocketBase

### DISCUSSION THREAD

**Discussion by bytekai** [pocketbase/pocketbase#4295] [Show and tell]
I've recently been experimenting with a UI based workflow system for https://pocketspace.dev/ to make it easier/faster to add custom functionality, I dont want to trap my users into my platform so a huge part of the design is making it possible for people to eject from pocketspace and self host their instances.

Currently there isn't any kind of escaping being done, so injection could happen but this is just a prototype.

This is what i have so far, i would love some feedback. The general idea is the json will be generated with a UI.

```ts
const newSignUpWorkflow = createWorkflow({
  triggerType: "onCollectionChange",
  type: "After",
  on: "Create",
  collection: "users",
  children: [
    {
      actionType: "sendEmail",
      to: "{{record.email}}",
      subject: "Welcome to our platform",
      body: "Welcome to our platform",
    },
  ],
});
```
Output
```js
onRecordAfterCreateRequest((e) => {

  const lt8kej = new MailerMessage({
    from: {
        address: $app.settings().meta.senderAddress,
        name:    $app.settings().meta.senderName,
    },
    to:      [{address: `${e.record.email}`}],
    subject: `Welcome to our platform`,
    html:    `Welcome to our platform`,
  })

  $app.newMailClient().send(lt8kej)

}, "users")
```

**Comment by ganigeorgiev** [comment]
I think it is a good idea (_I even have it in my never ending todo to implement something similar as an optional plugin once the Admin UI is refactored_) but note that there will be some breaking changes to the internals in order to simplify the exposed APIs and minimize the ambiguities/caveats.
I know that the lack of certainty probably doesn't help much but I'm still trying to figure out the best course of action before locking the APIs for v1, so just keep that in mind while working on your project to avoid spending too much time on something that most likely will change in the near future.

**Reply by bytekai** [reply]
Thank you for taking the time to write this, I will definitively keep this in mind. I'm not expecting anything to be stable until 1.0.

**Comment by bytekai** [comment]
Incase anybody reading this is curious. I've gotten pretty far. I've kept what gani said in mind and designed it to its easily refactorable when the eventual breaking changes accur.

![image](https://github.com/pocketbase/pocketbase/assets/73543613/42ebcac1-92dc-4fa3-b719-9a077b139a3e)

The majority of the plumbing is essentially done, its a bit rough on the edges, there is still alot of quality of life functionality i want to implement but its looking really good.
